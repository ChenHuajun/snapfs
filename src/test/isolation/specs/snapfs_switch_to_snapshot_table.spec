# Table testing.

setup {
	create table foo( id serial not null primary key, name varchar );
}

teardown {
	select pg_switch_to_snapshot( 0 );
	select pg_remove_snapshot( ( select recent_snapshot from pg_control_snapshot() ) );
	drop table foo;
}

session "s1"
step "s1_mk_sn" {
	select 1 as "make_snapshot" from pg_make_snapshot();
}

step "s1_sw_sn_1" {
	select pg_switch_to_snapshot( ( select recent_snapshot - 1 from pg_control_snapshot() ) );
}

step "s1_sw_sn_2" {
	select pg_switch_to_snapshot( ( select recent_snapshot from pg_control_snapshot() ) );
}

step "s1_sw_sn_0" {
	select pg_switch_to_snapshot( 0 );
}

step "s1_tab_i" {
	insert into foo ( name ) values ( 'some_text_sess_1' );
}

step "s1_tab_s" {
	select * from foo;
}

session "s2"
step "s2_tab_i" {
	insert into foo ( name ) values ( 'some_text_sess_2' );
}

step "s2_tab_u" {
	update foo set name = name || '_upd_sess_2';
}

step "s2_tab_s" {
	select * from foo;
}

permutation "s1_mk_sn" "s1_tab_i" "s2_tab_i" "s1_mk_sn" "s2_tab_u" "s1_tab_s" "s2_tab_s" "s1_sw_sn_1" "s1_tab_s" "s2_tab_s" "s1_sw_sn_2" "s2_tab_u" "s1_tab_s" "s2_tab_s" "s1_sw_sn_0" "s2_tab_u" "s1_tab_s" "s2_tab_s"
