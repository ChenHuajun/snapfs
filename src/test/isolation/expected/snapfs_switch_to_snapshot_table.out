Parsed test spec with 2 sessions

starting permutation: s1_mk_sn s1_tab_i s2_tab_i s1_mk_sn s2_tab_u s1_tab_s s2_tab_s s1_sw_sn_1 s1_tab_s s2_tab_s s1_sw_sn_2 s2_tab_u s1_tab_s s2_tab_s s1_sw_sn_0 s2_tab_u s1_tab_s s2_tab_s
step s1_mk_sn: 
	select 1 as "make_snapshot" from pg_make_snapshot();

make_snapshot  

1              
step s1_tab_i: 
	insert into foo ( name ) values ( 'some_text_sess_1' );

step s2_tab_i: 
	insert into foo ( name ) values ( 'some_text_sess_2' );

step s1_mk_sn: 
	select 1 as "make_snapshot" from pg_make_snapshot();

make_snapshot  

1              
step s2_tab_u: 
	update foo set name = name || '_upd_sess_2';

step s1_tab_s: 
	select * from foo order by id;

id             name           

1              some_text_sess_1_upd_sess_2
2              some_text_sess_2_upd_sess_2
step s2_tab_s: 
	select * from foo order by id;

id             name           

1              some_text_sess_1_upd_sess_2
2              some_text_sess_2_upd_sess_2
step s1_sw_sn_1: 
	select pg_switch_to_snapshot( ( select recent_snapshot - 1 from pg_control_snapshot() ) );

pg_switch_to_snapshot

               
step s1_tab_s: 
	select * from foo order by id;

id             name           

step s2_tab_s: 
	select * from foo order by id;

id             name           

step s1_sw_sn_2: 
	select pg_switch_to_snapshot( ( select recent_snapshot from pg_control_snapshot() ) );

pg_switch_to_snapshot

               
step s2_tab_u: 
	update foo set name = name || '_upd_sess_2';

ERROR:  Updates are prohibited in snapshot
step s1_tab_s: 
	select * from foo order by id;

id             name           

1              some_text_sess_1
2              some_text_sess_2
step s2_tab_s: 
	select * from foo order by id;

id             name           

1              some_text_sess_1
2              some_text_sess_2
step s1_sw_sn_0: 
	select pg_switch_to_snapshot( 0 );

pg_switch_to_snapshot

               
step s2_tab_u: 
	update foo set name = name || '_upd_sess_2';

step s1_tab_s: 
	select * from foo order by id;

id             name           

1              some_text_sess_1_upd_sess_2_upd_sess_2
2              some_text_sess_2_upd_sess_2_upd_sess_2
step s2_tab_s: 
	select * from foo order by id;

id             name           

1              some_text_sess_1_upd_sess_2_upd_sess_2
2              some_text_sess_2_upd_sess_2_upd_sess_2
